cmake/modules/FindPNETCDF.cmake

```diff
@@ -0,0 +1,73 @@
+# FindPNETCDF
+# -----------
+#
+# Find parallel IO library for classic netCDF files
+#
+# This module will define the following variables:
+#
+# ::
+#
+#   PNETCDF_FOUND
+#   PNETCDF_INCLUDE_DIRS
+#   PNETCDF_LIBRARIES
+#   PNETCDF_VERSION
+#   PNETCDF_ERANGE_FILL  - True if PnetCDF was built with ``NC_ERANGE`` support`
+#
+# The ``PNETCDF::pnetcdf`` target will also be exported
+
+include(FindPackageHandleStandardArgs)
+
+find_path(PNETCDF_INCLUDE_DIR NAMES pnetcdf.h)
+find_library(PNETCDF_LIBRARY NAMES pnetcdf)
+
+if (PNETCDF_INCLUDE_DIR)
+  set(pnetcdf_h "${PNETCDF_INCLUDE_DIR}/pnetcdf.h" )
+  message(STATUS "PnetCDF include file ${pnetcdf_h} will be searched for version")
+  file(STRINGS "${pnetcdf_h}" pnetcdf_major_string REGEX "^#define PNETCDF_VERSION_MAJOR")
+  string(REGEX REPLACE "[^0-9]" "" pnetcdf_major "${pnetcdf_major_string}")
+  file(STRINGS "${pnetcdf_h}" pnetcdf_minor_string REGEX "^#define PNETCDF_VERSION_MINOR")
+  string(REGEX REPLACE "[^0-9]" "" pnetcdf_minor "${pnetcdf_minor_string}")
+  file(STRINGS "${pnetcdf_h}" pnetcdf_sub_string REGEX "^#define PNETCDF_VERSION_SUB")
+  string(REGEX REPLACE "[^0-9]" "" pnetcdf_sub "${pnetcdf_sub_string}")
+  set(pnetcdf_version "${pnetcdf_major}.${pnetcdf_minor}.${pnetcdf_sub}")
+  message(STATUS "Found PnetCDF version ${pnetcdf_version}")
+endif()
+
+if (PNETCDF_DEBUG)
+  message(STATUS "[ ${CMAKE_CURRENT_LIST_FILE}:${CMAKE_CURRENT_LIST_LINE} ]"
+     " PNETCDF_ROOT = ${PNETCDF_ROOT}"
+    " PNETCDF_INCLUDE_DIR = ${PNETCDF_INCLUDE_DIR}"
+    " PNETCDF_LIBRARY = ${PNETCDF_LIBRARY}"
+    " PNETCDF_VERSION = ${PNETCDF_VERSION}"
+    )
+endif()
+
+file(STRINGS "${pnetcdf_h}" enable_erange_fill_pnetcdf REGEX "^#define PNETCDF_ERANGE_FILL")
+string(REGEX REPLACE "[^0-9]" "" erange_fill_pnetcdf "${enable_erange_fill_pnetcdf}")
+if("${erange_fill_pnetcdf}" STREQUAL "1")
+  set(PNETCDF_ERANGE_FILL ON CACHE BOOL "")
+else()
+  set(PNETCDF_ERANGE_FILL OFF CACHE BOOL "")
+endif()
+
+mark_as_advanced(PNETCDF_INCLUDE_DIR PNETCDF_LIBRARY PNETCDF_ERANGE_FILL)
+
+file(STRINGS "${pnetcdf_h}" relax_coord_bound_pnetcdf REGEX "^#define PNETCDF_RELAX_COORD_BOUND")
+string(REGEX REPLACE "[^0-9]" "" relax_coord_bound "${relax_coord_bound_pnetcdf}")
+# pnetcdf must have relaxed coord bounds to build with netCDF-4
+if(NOT "${relax_coord_bound}" STREQUAL "1")
+  message(FATAL_ERROR "Pnetcdf must be built with relax-coord-bound enabled")
```

Comment:
Generally it is bad to add `message(FATAL_ERROR)` within a `Find<Package>`. I am not sure about `message(WARNING)`

Just set `PNETCDF_FOUND = False` and `return`

Reply:
I think it's important that there's an informative message for the user here. Happy to make this change, but I'm not sure I see much difference between the methods of erroring here?

Reply:
The issue is when the module is called via `find_package(QUIET)` and/or without `REQUIRED`. For the former, the messages should not be displayed, but I am not sure up to which level does CMake suppress messages within `find_package(MODULE)` if at all. There is the `<PackageName>_FIND_QUIETLY` variable you can use of course. As for the using without `REQUIRED`, the issue is that `FATAL_ERROR` is that it will break the parent CMake script as well.

Reply:
Ah ok, I see your point. This isn't a general purpose module though, and we only use it with `find_package(REQUIRED)`, so I think overall it's probably nicer to have this check wrapped up in this module and `FATAL_ERROR` if there's an issue.

I guess if it were to be a general purpose module, this could set a feature variable instead, `PNETCDF_HAS_RELAXED_COORD_BOUNDS` or similar

Reply:
Indeed, it's just that I have seen many projects reuse `Find<Package>.cmake` from various projects, and these kind of issues just propagate. I generally document the modules to warn people not to use without consultation ([example](https://gitlab.com/octopus-code/octopus/-/blob/main/cmake/FindMETIS.cmake))